### ISSUE TITLE
clang++ 9 with libc++ fails to compile parsing.cc

### ISSUE THREAD

**Issue by DavidTruby** [flang-compiler/f18#664]
libc++ rL363692 onwards
This is due to a change in the standard meaning that narrowing conversions are no longer considered when selecting a constructor for variant. See http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0608r3.html
Unfortunately since this is a 'bug fix' in the standard it's likely this will also break in the same way with libstdc++ in future.
```
clang++ -stdlib=libc++  -I/home/davtru01/f18/build/../include -I/home/davtru01/f18/build/include -I/home/davtru01/llvm/include -std=c++17 -Wall -pedantic -fno-rtti -fno-exceptions -O2   -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -fsyntax-only ../lib/parser/parsing.cc
In file included from ../lib/parser/parsing.cc:15:
In file included from ../lib/parser/parsing.h:20:
In file included from ../lib/parser/instrumented-parser.h:19:
In file included from ../lib/parser/parse-state.h:28:
In file included from ../lib/parser/user-state.h:25:
../lib/parser/parse-tree.h:615:3: error: no matching constructor for initialization of 'std::variant<ScalarIntExpr, Star, Deferred>' (aka 'variant<Scalar<Integer<common::Indirection<Expr> > >, Fortran::parser::Star, Fortran::parser::TypeParamValue::Deferred>')
  UNION_CLASS_BOILERPLATE(TypeParamValue);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../lib/parser/parse-tree.h:105:22: note: expanded from macro 'UNION_CLASS_BOILERPLATE'
  classname(A &&x) : u(std::move(x)) {} \
                     ^ ~~~~~~~~~~~~
/home/davtru01/llvm/bin/../include/c++/v1/variant:620:9: note: in instantiation of function template specialization 'Fortran::parser::TypeParamValue::TypeParamValue<unsigned long, void>' requested here
      : __value(_VSTD::forward<_Args>(__args)...) {}
        ^
../lib/parser/basic-parsers.h:724:18: note: in instantiation of function template specialization 'Fortran::parser::CharLength::CharLength<unsigned long, void>' requested here
          return RESULT{std::move(*arg)};
                 ^
../lib/parser/basic-parsers.h:312:31: note: in instantiation of member function 'Fortran::parser::ApplyConstructor<Fortran::parser::CharLength, Fortran::parser::SequenceParser<Fortran::parser::Space, Fortran::parser::FollowParser<Fortran::parser::DigitString64, Fortran::parser::SpaceCheck> > >::Parse' requested here
    result = std::get<J>(ps_).Parse(state);
                              ^
../lib/parser/basic-parsers.h:299:9: note: in instantiation of function template specialization 'Fortran::parser::AlternativesParser<Fortran::parser::ApplyConstructor<Fortran::parser::CharLength, Fortran::parser::SequenceParser<Fortran::parser::TokenStringMatch<false, false>, Fortran::parser::FollowParser<Fortran::parser::Parser<Fortran::parser::TypeParamValue>, Fortran::parser::TokenStringMatch<false, false> > > >, Fortran::parser::ApplyConstructor<Fortran::parser::CharLength, Fortran::parser::SequenceParser<Fortran::parser::Space, Fortran::parser::FollowParser<Fortran::parser::DigitString64, Fortran::parser::SpaceCheck> > > >::ParseRest<1>' requested here
        ParseRest<1>(result, state, backtrack);
        ^
../lib/parser/grammar.h:653:1: note: in instantiation of member function 'Fortran::parser::AlternativesParser<Fortran::parser::ApplyConstructor<Fortran::parser::CharLength, Fortran::parser::SequenceParser<Fortran::parser::TokenStringMatch<false, false>, Fortran::parser::FollowParser<Fortran::parser::Parser<Fortran::parser::TypeParamValue>, Fortran::parser::TokenStringMatch<false, false> > > >, Fortran::parser::ApplyConstructor<Fortran::parser::CharLength, Fortran::parser::SequenceParser<Fortran::parser::Space, Fortran::parser::FollowParser<Fortran::parser::DigitString64, Fortran::parser::SpaceCheck> > > >::Parse' requested here
TYPE_PARSER(construct<CharLength>(parenthesized(typeParamValue)) ||
^
../lib/parser/type-parsers.h:47:19: note: expanded from macro 'TYPE_PARSER'
    return parser.Parse(state); \
                  ^
/home/davtru01/llvm/bin/../include/c++/v1/variant:1170:3: note: candidate constructor not viable: no known conversion from 'typename remove_reference<unsigned long &>::type' (aka 'unsigned long') to 'const std::__1::variant<Fortran::parser::Scalar<Fortran::parser::Integer<Fortran::common::Indirection<Fortran::parser::Expr, false> > >, Fortran::parser::Star, Fortran::parser::TypeParamValue::Deferred>' for 1st argument
  variant(const variant&) = default;
  ^
/home/davtru01/llvm/bin/../include/c++/v1/variant:1171:3: note: candidate constructor not viable: no known conversion from 'typename remove_reference<unsigned long &>::type' (aka 'unsigned long') to 'std::__1::variant<Fortran::parser::Scalar<Fortran::parser::Integer<Fortran::common::Indirection<Fortran::parser::Expr, false> > >, Fortran::parser::Star, Fortran::parser::TypeParamValue::Deferred>' for 1st argument
  variant(variant&&) = default;
  ^
/home/davtru01/llvm/bin/../include/c++/v1/variant:1183:13: note: candidate template ignored: substitution failure [with _Arg = unsigned long, $1 = 0, $2 = 0, $3 = 0]: no type named 'type' in 'std::__1::invoke_result<std::__1::__variant_detail::__overload<Fortran::parser::Scalar<Fortran::parser::Integer<Fortran::common::Indirection<Fortran::parser::Expr, false> > >, Fortran::parser::Star, Fortran::parser::TypeParamValue::Deferred>, unsigned long, unsigned long>'
  constexpr variant(_Arg&& __arg) noexcept(
            ^
/home/davtru01/llvm/bin/../include/c++/v1/variant:1192:22: note: candidate template ignored: could not match 'in_place_index_t<_Idx>' against 'unsigned long'
  explicit constexpr variant(
                     ^
/home/davtru01/llvm/bin/../include/c++/v1/variant:1220:22: note: candidate template ignored: could not match 'in_place_type_t<type-parameter-0-0>' against 'unsigned long'
  explicit constexpr variant(in_place_type_t<_Tp>, _Args&&... __args) noexcept(
                     ^
/home/davtru01/llvm/bin/../include/c++/v1/variant:1167:13: note: candidate constructor template not viable: requires 0 arguments, but 1 was provided
  constexpr variant() noexcept(is_nothrow_default_constructible_v<__first_type>)
            ^
/home/davtru01/llvm/bin/../include/c++/v1/variant:1206:22: note: candidate constructor template not viable: requires at least 2 arguments, but 1 was provided
  explicit constexpr variant(
                     ^
/home/davtru01/llvm/bin/../include/c++/v1/variant:1233:22: note: candidate constructor template not viable: requires at least 2 arguments, but 1 was provided
  explicit constexpr variant(
```

**Comment by DavidTruby** [comment]
I've also reproduced this when using gcc 10 and when using clang with libstdc++ from gcc 10.

**Comment by kiranchandramohan** [comment]
@DavidTruby has a fix for this. He will be making a PR soon.

**Comment by DavidTruby** [comment]
Fixed by #681
